rts/motoko-rts/src/types.rs

```diff
@@ -326,7 +329,7 @@ impl Value {
     /// Get the pointer as `Array` using forwarding. In debug mode panics if the value is not a pointer or the
     /// pointed object is not an `Array`.
     pub unsafe fn as_array(self) -> *mut Array {
-        debug_assert_eq!(self.tag(), TAG_ARRAY);
+        debug_assert!(self.tag() == TAG_ARRAY || self.tag() >= TAG_ARRAY_SLICE_MIN);
```

Comment:
Since we are relaxing this condition, we need to be careful to adjust all places that assume that arrays have `TAG_ARRAY` tags. @crusso: please keep this in mind when reviewing code.